packages/runtime/runtime-utils/src/serializer.ts

```diff
@@ -27,103 +130,38 @@ export class FluidSerializer implements IFluidSerializer {
 
     public get IFluidSerializer() { return this; }
 
-    public replaceHandles(
-        input: any,
-        bind: IFluidHandle,
-    ) {
+    /**
+     * Given a mostly-jsonable object that may have handle objects embedded within, will return a fully-jsonable object
+     * where any embedded IFluidHandles have been replaced with a serializable form.
+     *
+     * The original `input` object is not mutated.  This method will shallowly clones all objects in the path from
+     * the root to any replaced handles.  (If no handles are found, returns the original object.)
+     */
+    public replaceHandles(input: any, bind: IFluidHandle) {
         // If the given 'input' cannot contain handles, return it immediately.  Otherwise,
-        // return the result of 'recursivelyReplaceHandles()'.
-        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions,@typescript-eslint/no-unsafe-return
+        // return the result of recursively replacing handles with their encoded form.
+        // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
         return !!input && typeof input === "object"
-            ? this.recursivelyReplaceHandles(input, bind)
+            ? recursivelyReplace(bind, input, encodeValue)
             : input;
     }
 
     public stringify(input: any, bind: IFluidHandle) {
-        return JSON.stringify(input, (key, value) => {
-            // If the current 'value' is not a handle, return it unmodified.  Otherwise,
-            // return the result of 'serializeHandle'.
-            // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
-            const handle = !!value && value.IFluidHandle;
-            // TODO - understand why handle === false in some of our tests
-            // eslint-disable-next-line max-len
-            // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions,@typescript-eslint/no-unsafe-return
-            return handle
-                ? this.serializeHandle(handle, bind)
-                : value;
-        });
+        return JSON.stringify(input,
+            (key, value) =>
+                // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+                value && encodeValue(bind, value));
     }
```

Comment:
Make 'stringify()' use the same 'encodeValue()' function as 'replaceHandles()'...  just to make sure they are always in sync.